evm/src/memory/memory_stark.rs

```diff
@@ -409,15 +434,21 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for MemoryStark<F
             .constraint_transition(virtual_first_change * (next_addr_context - addr_context));
         yield_constr
             .constraint_transition(virtual_first_change * (next_addr_segment - addr_segment));
-        yield_constr.constraint_transition(address_unchanged * (next_addr_context - addr_context));
-        yield_constr.constraint_transition(address_unchanged * (next_addr_segment - addr_segment));
-        yield_constr.constraint_transition(address_unchanged * (next_addr_virtual - addr_virtual));
+        yield_constr.constraint_transition(
```

Comment:
Hmm I'm not sure if this is safe, it seems like our ordering could be "broken" by some invalid rows in between valid rows. E.g.

- valid row, time=6
- invalid row, time=8
- invalid row, time=2
- valid row, time=4

Maybe there's a way to make padding rows satisfy the regular constraints, e.g. we could record the last "real" memory timestamp, and add padding `MemoryOp`s at the end which start off at that timestamp. The padding `MemoryOp`s could perform some read that we know will succeed, e.g. repeating the last real read, or reading just beyond the last used address, which must be zero by default.

Or if not, we could add constraints to force padding rows to be together in a contiguous block, either at the beginning or the end.

In any case this seems nontrivial, so we could add a TODO for later.

Reply:
Hm well for now I think the only padding rows are entirely 0 values, except for the COUNTER column (and the COUNTER_PERMUTED and RANGE_CHECK_PERMUTED), and they're all at the beginning!

Reply:
Yeah our prover code puts them at the beginning, but seems like it'd be possible for a malicious prover to put them in the middle? If we wanted to enforce it we could constraint `is_valid` to have deltas in `{0, 1}`

Reply:
Ohh right

Reply:
Yeah that sounds great, thanks!